Delete all user forks tracked in RRefContext before shutting down

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

xush6528 committed:
Add user forks tracker in RRefContext

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -172,6 +172,7 @@ class RRefContext {
   //     in this map. A message of type RREF_USER_ACCEPT will remove the
   //     corresponding RRef from this map.
   std::unordered_map<ForkId, std::shared_ptr<RRef>, ForkId::Hash> pendingUsers_;
+  std::unordered_map<ForkId, std::weak_ptr<RRef>, ForkId::Hash> users_;
```
Add comments to describe the reason for adding this map?

xush6528 committed:
Update on "Add user forks tracker in RRefContext"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -354,6 +354,11 @@ void RRefContext::delPendingUser(const ForkId& forkId) {
   TORCH_INTERNAL_ASSERT(
       iter != pendingUsers_.end(),
       "Inconsistent states: attempt to delete a non-exist UserRRef.");
+  users_.emplace(
```
Why inserting it here instead of in `createUserRRef`?

xush6528 replied:
@mrshenli 

I think we should store confirmed user rrefs. Do you have a reason for tracking un-confirmed user rrefs?

mrshenli replied:
Makes sense, as only confirmed RRefs need to send delete message (it might be different when we add fault tolerant later). In that case, shall we call the map `confirmedUsers_` to be explicit?

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit c18290ad:


* **1/1** failures introduced in this PR


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

rohan-varma commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -109,20 +109,28 @@ UserRRef<T>::UserRRef(
-               << "RRefId = " << rrefId_ << ", ForkId = " << forkId_ << " : "
-               << "unknown error";
+void UserRRef<T>::tryDel() {
+  bool valid = valid_.exchange(false);
```
is it possible to call `tryDel` multiple times, hence this flag? In what cases could that happen?

xush6528 replied:
One is on destroy RRefContext, all UserRRefs will have their `tryDel()` called.

The other is when Python references a UserRRef, the destructor of UserRRef will be called. The destructor calls `tryDel`.

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP, debugging] Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -172,6 +174,7 @@ class RRefContext {
   //     in this map. A message of type RREF_USER_ACCEPT will remove the
   //     corresponding RRef from this map.
   std::unordered_map<ForkId, std::shared_ptr<RRef>, ForkId::Hash> pendingUsers_;
+  std::unordered_map<ForkId, std::weak_ptr<RRef>, ForkId::Hash> confirmedUsers_;
```
Add a comment about `confirmedUsers_`. Also, note that the comment above is now no longer valid as it says "The following two maps", but is actually referring to pendingUsers_ and pendingChildren_.

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1176,7 +1176,11 @@ def test_call_method_on_rref(self):
+    # through `torch.distributed.rpc.api`, so patching
+    # `torch.distributed.rpc._delete_all_user_rrefs` will not help.
+    @mock.patch.object(torch.distributed.rpc.api, "_delete_all_user_rrefs")
+    def _test_rref_leak(self, mock_delete_all_user_rrefs, ignore_leak):
```
`mock_delete_all_user_rrefs` isn't used anywhere in this PR?

xush6528 replied:
Right, the argument is added by the patch decorator. I can rename it to start with understore to indicate that it's a var name not used.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -257,6 +261,9 @@ class UserRRef final : public RRef {
   const ForkId forkId_;
+
+  // Indicates if this fork has sent delete message to it's owner.
+  std::atomic<bool> valid_{true};
```
nit: should rename this to a more appropriate name (ex: sentDelFork and default to false).

xush6528 replied:
No problem. That's definitely more readable.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -109,20 +109,28 @@ UserRRef<T>::UserRRef(
-               << "RRefId = " << rrefId_ << ", ForkId = " << forkId_ << " : "
-               << "unknown error";
+void UserRRef<T>::tryDel() {
+  bool valid = valid_.exchange(false);
```
This logic is not correct, we are setting valid to false before we actually send out the delFork message. We should do this after sending the delFork message. It could be possible that the delFork fails, but we have set valid to false, which is incorrect.

xush6528 replied:
@pritamdamania87 

Nice catch.

Notice that, that way, we might send out the delete user rref request twice.

One by the Python GC, the other by the shutdown call.

So I need to remove the assertion that limits this request to once too.

pritamdamania87 reviewed (changes requested):
Thanks for fixing this! I have a few comments inline.

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -207,6 +207,8 @@ class RRef : public RRefInterface {
     return rrefId_;
   }
 
+  virtual void tryDel() {}
```
ideally, this should only live in `UserRRef`. This is OK for now, as otherwise we would have to do static_cast. After #30630, we should be able to let `confirmedUsers_` store `UserRRef`, and can then delete this from the base class.

xush6528 replied:
This was my concern too. Thanks for telling me about #30630.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -207,6 +208,7 @@ def shutdown(graceful=True):
     if graceful:
         _wait_all_workers()
+        _delete_all_user_rrefs()
         _agent.join()
```
If `ThriftAgent.join()` is no-op, how will this work for that?

xush6528 replied:
This PR will not help ThriftAgent unfortunately, as I have tracked this situation in https://github.com/pytorch/pytorch/issues/31325#issuecomment-574918917.
I must update `_wait_all_workers()` as the next step.

mrshenli reviewed (commented):
Test failures are real.

```
Jan 16 01:22:10 ======================================================================
Jan 16 01:22:10 FAIL [1.326s]: test_nested_rref (__main__.RpcTestWithSpawn)
Jan 16 01:22:10 ----------------------------------------------------------------------
Jan 16 01:22:10 Traceback (most recent call last):
Jan 16 01:22:10   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 130, in wrapper
Jan 16 01:22:10     self._join_processes(fn)
Jan 16 01:22:10   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 211, in _join_processes
Jan 16 01:22:10     self._check_return_codes(elapsed_time)
Jan 16 01:22:10   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 231, in _check_return_codes
Jan 16 01:22:10     self.assertEqual(p.exitcode, first_process.exitcode)
Jan 16 01:22:10   File "/var/lib/jenkins/workspace/test/common_utils.py", line 890, in assertEqual
Jan 16 01:22:10     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
Jan 16 01:22:10 AssertionError: 7 not less than or equal to 1e-05 : 
```

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -164,16 +164,18 @@ class RRefContext {
+  //     corresponding RRef from pendingUsers_ map to confirmedUsers_ map.
   std::unordered_map<ForkId, std::shared_ptr<RRef>, ForkId::Hash> pendingUsers_;
+  //     confirmedUsers_ map is useful for proactively delete UserRRefs
+  //     on shutting down RPC stack, avoiding OwnerRRef leaks.
```
Let's add: UserRRefs are added into this map when it is confirmed by the owner. When destroying RRefContext this map helps to find local UserRRefs and send delete messages if they are still alive.

xush6528 replied:
Updated.

mrshenli reviewed (commented):
`test_nested_rref_stress` has failed. Seems one worker quits too soon.

```
Jan 16 05:38:51 ......s...................terminate called after throwing an instance of 'gloo::IoException'
Jan 16 05:38:51   what():  [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:563] Read error [172.17.0.2]:59041: Connection reset by peer
Jan 16 05:38:51 terminate called after throwing an instance of 'gloo::IoException'
Jan 16 05:38:51   what():  [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:563] Read error [172.17.0.2]:32406: Connection reset by peer
Jan 16 05:38:56 ERROR:torch.distributed.rpc.api:worker2 failed to respond to 'Shutdown Proceed.' request in 0:00:05
Jan 16 05:38:56 ERROR:torch.distributed.rpc.api:worker3 failed to respond to 'Shutdown Proceed.' request in 0:00:05
```

xush6528 committed:
Update on "[WIP, debugging] Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP, debugging] Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 commented:
"pytorch_linux_xenial_py3_6_gcc5_4_test" is a flaky test project. It does not depends on cuda. Trying to reproduce it in Docker on my MacOS.

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -207,6 +208,7 @@ def shutdown(graceful=True):
 
     if graceful:
         _wait_all_workers()
+        _delete_all_user_rrefs()
```
I am trying to figure out why the test is flaky on this PR. 

```
Jan 17 02:11:26 Traceback (most recent call last):
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
Jan 17 02:11:26     self.run()
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
Jan 17 02:11:26     self._target(*self._args, **self._kwargs)
Jan 17 02:11:26   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 198, in _run
Jan 17 02:11:26     getattr(self, test_name)()
Jan 17 02:11:26   File "/var/lib/jenkins/workspace/test/dist_utils.py", line 75, in new_test_method
Jan 17 02:11:26     rpc.shutdown(graceful=clean_shutdown)
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/rpc/api.py", line 69, in wrapper
Jan 17 02:11:26     return func(*args, **kwargs)
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/rpc/api.py", line 212, in shutdown
Jan 17 02:11:26     _agent.join()
Jan 17 02:11:26 RuntimeError: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [172.17.0.2]:51896
```

The above error suggests that allgather failed because some peer exits too early. There are two possibilities:

1. some peer leaves join() too earlier
2. some peer crashed. 

1 should not happen because, allgather guarantees that all peers have the same view on the global snapshot, so that they leave the join() function after the same amount of iterations. It's impossible (unless there is a bug in join) that one peer wants to do one more round than another peer.

So the problem is some peer crashed somewhere. Will the following case be possible?

In the `test_nested_rref_stress`, we have the following code:

```python
        for i in range(20):
            rref_of_rrefs = all_rrefs[i]
            rrefs = rref_of_rrefs.to_here()
```

Let's assume this is called on rank 0. It means that rank0 gets a fork/UserRRef from rank1's UserRRef. So, according to our protocol, rank1 will hold it's RRef alive until rank0's UserRRef is confirmed by the Owner, preventing the Owner being deleted too soon. However, it is possible that, after `_wait_all_workers()`, rank0 still didn't get the confirmation from the Owner, and `_delete_all_user_rrefs` deletes the UserRRef on rank1, which might trigger the OwnerRRef been deleted before knowing that there is a fork on rank0. 

If the above is true, then, the fix would be calling `ProcessGroupAgent::sync()` before  `_delete_all_user_rrefs()` to make sure all messages are processed before deleting UserRRefs. The `sync` API was removed, but I guess we can expose it for testing purpose? If this is indeed the problem, we can then discuss whether we want it back, or can `_wait_all_workers()` do anything about it.

a holding rrf alive for c, but been forcely erased by _delete_all_user_rrefs() which trigger owner rrefs to die

mrshenli replied:
Let's see if test pass: #32398

xush6528 replied:
This PR is for removing sync() API, but the problem this PR triggers require us to have sync() API, so this is not the right solution. Closing.

xush6528 committed:
Update on "【WIP】Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "【WIP】Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "【WIP】Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "【WIP】Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "【WIP】Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"


In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -380,17 +411,24 @@ void RRefContext::addPendingUser(
-      iter != pendingUsers_.end(),
-      "Inconsistent states: attempt to delete a non-exist UserRRef.");
-  pendingUsers_.erase(iter);
+void RRefContext::confirmPendingUser(const ForkId& forkId) {
```
Any reason for making this name change? Its name collides with another `confirmPendingUser` in this file.

xush6528 replied:
Right, this is more confusing now. Let me revert back. I was thinking the name "delPendingUser” can't reflect the fact that it also adds to the `confimedUsers_` map.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -195,17 +196,24 @@ class TORCH_API RRefContext {
       forks_;
 
-  // The follow two maps keep UserRRefs alive by holding a intrusive_ptr to the
+  std::condition_variable pendingReduceCV_;
```
need comments on this CV. BTW, why there is a `Reduce` in the name?

xush6528 replied:
I am using mussel memory from that 2 cond vars on producer-consumer queue, one is for item added, the other for item removed. 

Since we are waiting on both sizes to be 0, I think only need to notify `pendingUsers_.erase(iter)` and `pendingChildren_.erase(iter)`. `pendingUsers_.empalce(**)` and `pendingChildren_.emplace(**)` don't need to notify this CV.

mrshenli reviewed (commented):
Do we also need to block on `RRefContext::destroyInstance` when OwnerRRefs are not empty? And this needs to happen before agent join. Otherwise, we could run into the following situation:

On node A:
It has no UserRRef, and has one OwnerRRef used by node B. Then A would skip `_delete_all_user_rrefs`, run `RpcAgnet.join()`, and then delete the context. 

On node B:
It will first run `_delete_all_user_rrefs`, and then run `RpcAgnet.join(). However, what could happen is that A could have finished `RpcAgent.join()` even before B reaches `_delete_all_user_rrefs`. And it will hit "closed by peers" error.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -235,6 +236,7 @@ def shutdown(graceful=True):
+        _delete_all_user_rrefs()
         _get_current_rpc_agent().join()
     try:
         # This raises a `TORCH_CHECK()` exception on RRef leak detected.
```
to be more precise, we need to move the line below (`_destroy_rref_context `) to the front of `_get_current_rpc_agent().join()`. And we probably can also consolidate `_destroy_rref_context` and `_delete_all_user_rrefs` to one function, e.g., `wait_and_clear_rref_context`.

xush6528 replied:
The problem is that I didn't wait for response of delete user rref.
So user could run to `_agent.join()` right after sending delete user rref request, while owner is also waiting on `_agent.join()`. At this time, owner can destroy rref context and report owner rref leaks. After this, the server thread picks up the request to delete user rref and realized no such rref id exists and the owner node crashes. Then, User node detects GLOO connection cloed by peer.

xush6528 replied:
Keeping it as is, cause we don't want to mix graceful and non-graceful together. Non-graceful should not wait on anything.

xush6528 commented:
> A could have finished RpcAgent.join() even before B reaches `_delete_all_user_rrefs`

How could it happen?

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"

Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -165,8 +165,38 @@ void RRefContext::delUser(
+  bool noPendingAndOwner =
+      pendingReducedCV_.wait_for(lock, timeoutMillis, [this]() {
+        return pendingUsers_.size() == 0 && pendingChildren_.size() == 0 &&
+            owners_.size() == 0;
```
don't we need to wait for owners after deleting all users? I mean sth like:

1. wait for pendingUsers_ and pendingChildren_
2. delete all users 
3. wait for all owners_?

Otherwise, A could be holding a user rref of B, while B could holding a user rref to A? and neither of them would proceed?

mrshenli replied:
I tried making that change, but could still hit the following error. Hmm...

```
Traceback (most recent call last):
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shenli/pytorch/torch/testing/_internal/common_distributed.py", line 243, in _run
    getattr(self, test_name)()
  File "/data/users/shenli/pytorch/torch/testing/_internal/dist_utils.py", line 71, in new_test_method
    rpc.shutdown(graceful=clean_shutdown)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 77, in wrapper
    return func(*args, **kwargs)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 240, in shutdown
    _get_current_rpc_agent().join()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [2401:db00:21:6058:face:0:3:0]:5529
ERROR:torch.distributed.rpc.api:worker2 failed to respond to 'Shutdown Proceed.' request in 0:00:05
Process process 0:
Traceback (most recent call last):
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shenli/pytorch/torch/testing/_internal/common_distributed.py", line 243, in _run
    getattr(self, test_name)()
  File "/data/users/shenli/pytorch/torch/testing/_internal/dist_utils.py", line 71, in new_test_method
    rpc.shutdown(graceful=clean_shutdown)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 77, in wrapper
    return func(*args, **kwargs)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 240, in shutdown
    _get_current_rpc_agent().join()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [2401:db00:21:6058:face:0:3:0]:38794
```

xush6528 replied:
The problem is that I didn't wait for response of delete user rref.
So user could run to `_agent.join()` right after sending delete user rref request, while owner is also waiting on `_agent.join()`. At this time, owner can destroy rref context and report owner rref leaks. After this, the server thread picks up the request to delete user rref and realized no such rref id exists and the owner node crashes. Then, User node detects GLOO connection cloed by peer.

xush6528 replied:
"wait for all owners_" should be interchangeable with "agent.join()", if client waits for response.

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"

Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"

Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 commented:
On this version, the error is 
```
[E rref_impl.cpp:91] Error occurred when deleting UserRRef instance, RRefId = GloballyUniqueId(0, 0), ForkId = GloballyUniqueId(0, 1) : responseMessageFuturePtr != nullptr INTERNAL ASSERT FAILED at /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_impl.cpp:87, please report a bug to PyTorch.  (tryDel at /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_impl.cpp:87)
```

The assertion does not hold.

mrshenli commented:
> On this version, the error is
>
> > [E rref_impl.cpp:91] Error occurred when deleting UserRRef instance, RRefId = GloballyUniqueId(0, 0), ForkId = GloballyUniqueId(0, 1) : responseMessageFuturePtr != nullptr INTERNAL ASSERT FAILED at /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_impl.cpp:87, please report a bug to PyTorch. (tryDel at /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_impl.cpp:87)
>
> The assertion does not hold.

This is weird. This would only happen when `deletedOnOwner_ == false` and `destroyed_==true`? It means the RRefContext has been destroyed but somehow there are still UserRRefs that have never called `tryDel`?

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"

Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

xush6528 commented:
merged the version in https://github.com/pytorch/pytorch/pull/34610, so the change compared to  last night are,

1) 
In RRefContext::delAllUserRRef(), copy confirmedUsers to a temp, so that we can release lock, so RRef->tryDel() and Context::delUser(..) can acquire lock, and remove the deleted UserRRef from confirmedUsers_.

A more frequent call path to RRef->tryDel() is Python GC triggered UserRRef destructor, where the wrapping logic does not acquire lock at all.

2) 
After sending all delete user RRef messages, should wait for number of OwnerRRefs to reach 0 before moving on.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -360,12 +405,15 @@ void RRefContext::addPendingChild(
+    TORCH_INTERNAL_ASSERT(
+        iter != pendingChildren_.end(),
+        "Inconsistent states: attempt to delete a non-exist child fork.");
+    pendingChildren_.erase(iter);
```
cannot directly erase the `UserRRef` here, because that could trigger `delUser` which would then acquire lock on `mutex_` to modify `confirmedUsers_`. As the lock is already acquired in this block, it would deadlock. 

#34610 solved this problem by creating a temporary `c10::intrusive_ptr<RRef> deletedUser` to keep the deleted UserRRef alive in this block.

xush6528 replied:
Thanks for explaining the subtlety here. Updating with comments.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -381,12 +429,19 @@ void RRefContext::addPendingUser(
+        std::piecewise_construct,
+        std::forward_as_tuple(forkId),
+        std::forward_as_tuple(iter->second));
+    pendingUsers_.erase(iter);
```
ditto

xush6528 replied:
Updating.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -157,16 +157,59 @@ void RRefContext::delUser(
+  // First, wait for all pending UserRRefs to be confirmed,
+  // one kind is pendingUsers_, which are shared from Owner,
+  // the other kind pendingChildren_, which are shared from another User.
+  std::unique_lock<std::mutex> lock(mutex_);
```
looks like there is nothing that need to be shared across the three lock-unlock-lock blocks, we can use two lock_guard here to make the code more readable?

xush6528 replied:
sounds good. updating.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -157,16 +157,59 @@ void RRefContext::delUser(
+  // Note, there should be no new forkings in between, because it's assumed that
+  // this utility is called during graceful shutdown, where no new user RPCs can
+  // not be initiaited anymore.
+  auto tempConfirmedUsers = confirmedUsers_;
```
as erasing non-existing keys from unordered_map is fine, we can swap here?

xush6528 replied:
Yeah, we can. It's more performant, like move semantics, but bi-directional. 
Also, actually no one is expected to insert to confirmedUsers_ again, based on the assumption that it's called during graceful shutdown.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -208,7 +251,7 @@ c10::intrusive_ptr<RRef> RRefContext::getOrCreateRRef(
     const RRefId& rrefId,
     const TypePtr& type) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::unique_lock<std::mutex> lock(mutex_);
```
This is one debugging vestige in #34610. I feel we don't need this in this PR. Let me test in #34610. If not necessary, let's revert these three lines.

xush6528 replied:
Reverting.

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -213,6 +213,8 @@ class TORCH_API RRef : public RRefInterface {
     return type_;
   }
 
+  virtual void tryDel() {}
```
Let's add comments to explain its behavior.

xush6528 replied:
Adding.

mrshenli reviewed (commented):
We are getting close. If tests pass after adding `c10::intrusive_ptr<RRef> deletedUser;`, we can land. :)

mrshenli reviewed (commented):
Could you please add a test for circular dependency? The example in #27096 should work.

xush6528 commented:
Update: Use temporary ref to delay UserRRef destruction until lock released.

xush6528 commented:
Added test case to show the "circular depenedncy" in #27096. 
To be precise, it's because Python GC doesn't collect local circular dependent object, unless explicitly call `gc.collect()`. `UserRRef`s held by such Python objects will not be deleted in time.

xush6528 committed:
Update on "Delete all user forks tracked in RRefContext before shutting down"

Stack from [ghstack](https://github.com/ezyang/ghstack):

* **#31893 Delete all user forks tracked in RRefContext before shutting down**

In order to resolve the issue summarized in https://github.com/pytorch/pytorch/issues/31325.

The overal solution is to proactively send out delete fork messages from user nodes, before user nodes detecting rref leaks.

As the first step, we want to have a weak ref tracker to track all user rrefs.

Differential Revision: [D19292254](https://our.internmc.facebook.com/intern/diff/D19292254/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -23,6 +23,8 @@ namespace rpc {
 
 namespace {
 
+constexpr std::chrono::milliseconds kDeleteAllUsersTimeout(100000);
```
This is looks good to me for now. In future PRs we probably would like to expose this to users for configure?

xush6528 replied:
Yes, for now I prefer not exposing to shutdown API this moment.

Because I think users should only care to pass a top-level timeout, that includes 
0) wait_for_all_workers, to flush any user RPC requests.
1) wait for pending, 
2) send and wait delete user forks, 
3) wait for owners to be removed. 
4) wait for join().

This will require `utils::Future` to take timeout on it's wait() function, and require RpcAgent sync()/join() to take timeout as well.

mrshenli reviewed (approved):
Awesome!! Finally, we got all tests passed. Thanks a lot for spending so much time on this PR. LGTM! Let's land!
